File: analytics/src/main/java/io/appmetrica/analytics/impl/utils/concurrency/FileLocksHolder.java
package io.appmetrica.analytics.impl.utils.concurrency;

import android.annotation.SuppressLint;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import java.util.HashMap;
import java.util.Map;

public class FileLocksHolder {

    @SuppressLint("StaticFieldLeak")
    @Nullable
    private static volatile FileLocksHolder INSTANCE;

    @NonNull
    private final Context context;
    private final Map<String, ExclusiveMultiProcessFileLock> locks =
            new HashMap<String, ExclusiveMultiProcessFileLock>();

    @NonNull
    public static FileLocksHolder getInstance(@NonNull Context context) {
        if (INSTANCE == null) {
            synchronized (FileLocksHolder.class) {
                if (INSTANCE == null) {
                    INSTANCE = new FileLocksHolder(context);
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    FileLocksHolder(@NonNull Context context) {
        this.context = context;
    }

    @NonNull
    public ExclusiveMultiProcessFileLock getOrCreate(@NonNull String fileName) {
        if (!locks.containsKey(fileName)) {
            synchronized (this) {
                if (!locks.containsKey(fileName)) {
                    locks.put(fileName, new ExclusiveMultiProcessFileLock(context, fileName));
                }
            }
        }
        return locks.get(fileName);
    }

    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    static void clearInstance() {
        INSTANCE = null;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/utils/concurrency/FileLocksHolder.kt
package io.appmetrica.analytics.impl.utils.concurrency

import android.annotation.SuppressLint
import android.content.Context
import androidx.annotation.VisibleForTesting

class FileLocksHolder @VisibleForTesting internal constructor(private val context: Context) {
    private val locks: MutableMap<String, ExclusiveMultiProcessFileLock> = HashMap()

    @Synchronized
    fun getOrCreate(fileName: String): ExclusiveMultiProcessFileLock {
        return locks.getOrPut(fileName) { ExclusiveMultiProcessFileLock(context, fileName) }
    }

    @Synchronized
    fun clear(fileName: String) {
        locks.remove(fileName)
    }

    companion object {

        @SuppressLint("StaticFieldLeak")
        @Volatile
        private lateinit var INSTANCE: FileLocksHolder

        @JvmStatic
        fun getInstance(context: Context): FileLocksHolder {
            if (!::INSTANCE.isInitialized) {
                synchronized(FileLocksHolder::class) {
                    if (!::INSTANCE.isInitialized) {
                        INSTANCE = FileLocksHolder(context)
                    }
                }
            }
            return INSTANCE
        }

        @VisibleForTesting
        fun stubInstance(value: FileLocksHolder) {
            INSTANCE = value
        }
    }
}
